### PR TITLE
New resource r/aws_wafv2_web_acl

### LINKED ISSUE #11176: Feature Request: WAFv2 Web ACL Resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In November 2019, AWS released a new version of the WAF API, [WAFv2](https://docs.aws.amazon.com/waf/latest/APIReference/Welcome.html), which offers improved functionality over the previous WAF API ("WAF Classic") such as [Managed Rules](https://github.com/terraform-providers/terraform-provider-aws/issues/11046) and WAF Capacity Units. This new API requires separate Terraform resource implementations from the previous resource implementations.

### New or Affected Resource(s)

* aws_wafv2_web_acl

### Potential Terraform Configuration

```hcl
resource "aws_wafv2_web_acl" "example" {
  default_action {
    allow {}
    block {}
  }

  description = ""
  name        = ""
  scope       = ""
  tags        = {}

  rule {
    action {
      # The below shows the current API structure, which means there may be intention to include
      # additional settings with each action type in the future
      # Purposefully empty configuration blocks are awkward in Terraform configurations and the Plugin SDK though, so we may want to shy away from the pure API schema mapping here if they cause any trouble
      allow {}
      block {}
      count {}
    }

    name = ""
    
    override_action {
      count {}
      none {}
    }

    priority = 1

    # these are not fully shown and appear to be recursive in some cases    
    statement {
      and_statement {
        statement {}
        statement {}
      }

      byte_match_statement {
        field_to_match {
          all_query_arguments {}
          body {}
          method {}
          query_string {}
          single_header {
            name = ""
          }
          single_query_argument {
            name = ""
          }
          uri_path {}
        }

        positional_constraint = ""
        search_string = ""

        text_transformation {
          priority = 1
          type = ""
        }
      }

      geo_match_statement {
        country_codes = []
      }

      ip_set_reference_statement {
        arn = ""
      }

      managed_rule_group_statement {
        excluded_rule {
          name = ""
        }

        name = ""
        vendor_name = ""
      }

      not_statement {
        statement {}
      }

      or_statement {
        statement {}
        statement {}
      }

      rate_based_statement {
        aggregate_key_type = ""
        limit = 100

        scope_down_statement {}
      }

      regex_pattern_set_reference_statement {
        arn = ""

        field_to_match {} # see above
        text_transformation {} # see above
      }

      rule_group_reference_statement {
        arn = ""

        excluded_rule {
          name = ""
        }
      }

      size_constraint_statement {
        comparison_operator = ""

        field_to_match {} # see above

        size = 1

        text_transformation {} # see above
      }

      sqli_match_statement {
        field_to_match {} # see above

        text_transformation {} # see above
      }

      xss_match_statement {
        field_to_match {} # see above

        text_transformation {} # see above
      }
    }
  }

  visibility_config {
    cloudwatch_metrics_enabled = true
    metric_name = ""
    sampled_requests_enabled = true
  }
}
```

### References

- Requires: https://github.com/terraform-providers/terraform-provider-aws/pull/11172
- API Reference: https://docs.aws.amazon.com/waf/latest/APIReference/API_Operations_AWS_WAFV2.html
- SDK Reference: https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/

### PR DESCRIPTION
**Pull request by pvanbuijtene**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11176
Relates #11046

This PR can not be merged before #12677 because of copied dependencies.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
```release-note
* **New Resource**: `aws_wafv2_web_acl` (#12688)
```

Output from acceptance testing:

```
make build; make testacc TEST=./aws TESTARGS='-run=TestAccAwsWafv2WebACL'
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsWafv2WebACL -timeout 120m
...
--- PASS: TestAccAwsWafv2WebACL_Disappears (83.89s)
--- PASS: TestAccAwsWafv2WebACL_Minimal (89.98s)
--- PASS: TestAccAwsWafv2WebACL_ChangeNameForceNew (172.23s)
--- PASS: TestAccAwsWafv2WebACL_ManagedRuleGroupStatement (191.75s)
--- PASS: TestAccAwsWafv2WebACL_RateBasedStatement (196.14s)
--- PASS: TestAccAwsWafv2WebACL_Basic (197.97s)
--- PASS: TestAccAwsWafv2WebACL_Tags (222.97s)
--- PASS: TestAccAwsWafv2WebACL_RuleGroupReferenceStatement (255.26s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       228.215s
```

Some of the code can be refactored after `aws_wafv2_rule_group` #12677 is merged.

### CODE REVIEW

**Comment by anGie44** [comment]
hi @pvanbuijtene 👋 -- dropping a note here to rebase and isolate the resource changes when you get a chance :)

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - The ARN of the WAF rule group.
+* `capacity` - The web ACL capacity units (WCUs) currently being used by this web ACL.
+* `id` - The ID of the WAF rule group.
+
+## Import
+
+WAFv2 Rule Group can be imported using the ID, Name and Scope e.g.
```
[the remark this answers is not in the document]
Comment (anGie44) [reply website/docs/r/wafv2_web_acl.html.markdown]: nice already resolved 😄

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 2]
```diff
@@ -0,0 +1,1128 @@
...
+					resource.TestCheckResourceAttr(resourceName, "default_action.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "default_action.0.allow.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "default_action.0.block.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.0.cloudwatch_metrics_enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.0.metric_name", "friendly-metric-name"),
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.0.sampled_requests_enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "2"),
+					resource.TestCheckResourceAttr(resourceName, "tags.Tag1", "Value1"),
+					resource.TestCheckResourceAttr(resourceName, "tags.Tag2", "Value2"),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:43]: let's also verify the starting expected `rule.#`

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 3]
```diff
@@ -0,0 +1,1128 @@
...
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateIdFunc: testAccAwsWafv2WebACLImportStateIdFunc(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2WebACL_Minimal(t *testing.T) {
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:228]: could we similarly add a test w/empty blocks where possible, i.e. `rule {}`?

File: aws/resource_aws_wafv2_web_acl.go [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				// Required field but currently only supports "IP"
+				"aggregate_key_type": {
+					Type:     schema.TypeString,
+					Optional: true,
+					Default:  "IP",
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl.go:400]: also available in the sdk as `wafv2.RateBasedStatementAggregateKeyTypeIp`

File: aws/resource_aws_wafv2_web_acl.go [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+				},
+				"limit": {
+					Type:         schema.TypeInt,
+					Required:     true,
+					ValidateFunc: validation.IntBetween(100, 2000000000.),
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl.go:405]: small typo: should be the int `2000000000` without the decimal

File: aws/resource_aws_wafv2_web_acl.go [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+					wafv2.ScopeCloudfront,
+					wafv2.ScopeRegional,
+				}, false),
+			},
+			"rule": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"action": {
```
[the remark this answers is not in the document]
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl.go:92]: @anGie44  There are a few rules which I didn't exactly know how to validate when working on this, so only focused on basic implementation. I can look into it if you want.

But I don't think `default_action` and `action` are mutual exclusive 🤔

File: aws/resource_aws_wafv2_web_acl_test.go (hunk 3 above)
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl_test.go:230]: Empty block is not possible as there are required fields, what would you like to test?

Comment (anGie44) [reply aws/resource_aws_wafv2_web_acl_test.go:230]: i overlooked the first level under `rule` was not fully optional..disregard please!

File: aws/resource_aws_wafv2_web_acl.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl.go:92]: I looked into some validation but think it's currently not possible because it's scoped on rules.

Would be nice to use `ValidateFunc` but I run into `ValidateFunc is not yet supported on lists or sets.`, so think there's not much possible currently :(

[the remark this answers is not in the document]
Comment (anGie44) [reply aws/resource_aws_wafv2_web_acl.go:92]: thanks for the research @pvanbuijtene! i guess I had in mind using something like `ExactlyOneOf` but it does seem a bit restrictive in case things change at the AWS API level for some reason...the schema is most likely fine as is, I just wanted to bring it up for discussion as I saw the docs pointed out the use of 1 of the 2 fields but not both

[the remark this answers is not in the document]
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl.go:92]: I also looked into `CustomizeDiff` but it looks like the the data for the `rules` is not available from the root, and the `CustomizeDiff` on the `rule` schema is not being called it looks like.

**Comment by anGie44** [comment]
hi @pvanbuijtene, thanks for adding the last code dependency. as it's a collection of helper functions used across this resource and the rule_group resource (and maybe unintentionally is missing the complete rule_group resource definition?), can we rename it to something like `wafv2_helpers.go` 

otherwise, atm this PR is looking good with output of acceptance tests:
```
--- FAIL: TestAccAwsWafv2WebACL_RuleGroupReferenceStatement (3.34s) -- expected error: The provider provider.aws does not support resource type "aws_wafv2_rule_group".
--- PASS: TestAccAwsWafv2WebACL_Disappears (40.84s)
--- PASS: TestAccAwsWafv2WebACL_Minimal (43.84s)
--- PASS: TestAccAwsWafv2WebACL_ChangeNameForceNew (99.09s)
--- PASS: TestAccAwsWafv2WebACL_ManagedRuleGroupStatement (112.34s)
--- PASS: TestAccAwsWafv2WebACL_Basic (128.26s)
--- PASS: TestAccAwsWafv2WebACL_RateBasedStatement (133.31s)
--- PASS: TestAccAwsWafv2WebACL_Tags (161.10s)
```

File: aws/resource_aws_wafv2_web_acl.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anGie44) [reply aws/resource_aws_wafv2_web_acl.go:92]: noted, thank you! then let's keep things as they are here 👍

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 7]
```diff
@@ -0,0 +1,1126 @@
...
+					"geo_match_statement":                   []interface{}{},
+					"ip_set_reference_statement":            []interface{}{},
+					"managed_rule_group_statement":          []interface{}{},
+					"not_statement":                         []interface{}{},
+					"or_statement":                          []interface{}{},
+					"rate_based_statement":                  []interface{}{},
+					"regex_pattern_set_reference_statement": []interface{}{},
+					"rule_group_reference_statement": []interface{}{
+						map[string]interface{}{
+							"arn":           *r.Rules[0].Statement.RuleGroupReferenceStatement.ARN,
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:472]: Nit: extract with `aws.StringValue()`

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+* `single_query_argument` - (Optional) Inspect a single query argument. See [Single Query Argument](#single-query-argument) below for details.
+* `uri_path` - (Optional) Inspect the request URI path. This is the part of a web request that identifies a resource, for example, `/images/daily-ad.jpg`.
+
+### Single Header
+
+Inspect a single header. Provide the name of the header to inspect, for example, `User-Agent` or `Referer`. This setting isn't case sensitive.
+
+The `single_header` block supports the following arguments:
+
+* `name` - (Optional) The name of the query header to inspect. This setting isn't case sensitive, only use lower case characters.
```
Comment (anGie44) [inline website/docs/r/wafv2_web_acl.html.markdown:456]: this line may be a bit contradictory (i.e. for `name` to not be case sensitive but only lower case). we can re-word this to say something like "This setting must be provided as lower case characters.")

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+* `body` - (Optional) Inspect the request body, which immediately follows the request headers.
+* `method` - (Optional) Inspect the HTTP method. The method indicates the type of operation that the request is asking the origin to perform.
+* `query_string` - (Optional) Inspect the query string. This is the part of a URL that appears after a `?` character, if any.
+* `single_header` - (Optional) Inspect a single header. See [Single Header](#single-header) below for details.
+* `single_query_argument` - (Optional) Inspect a single query argument. See [Single Query Argument](#single-query-argument) below for details.
+* `uri_path` - (Optional) Inspect the request URI path. This is the part of a web request that identifies a resource, for example, `/images/daily-ad.jpg`.
+
+### Single Header
+
+Inspect a single header. Provide the name of the header to inspect, for example, `User-Agent` or `Referer`. This setting isn't case sensitive.
```
Comment (anGie44) [inline website/docs/r/wafv2_web_acl.html.markdown:452]: these examples could be misleading to a reader w/o further reading of the bullet points below - to err on the side of caution, we could append parens with something like: "for example, `User-Agent` or `Referer`(provided as lowercase strings)."

**Comment by pvanbuijtene** [comment]
@anGie44 I renamed the file and can resolve the duplicate functions when the other PR is merged.

**Comment by anGie44** [comment]
hi again @pvanbuijtene 👋 . #12677 is now merged but the above conflicts have come up. mind rebasing when you have a chance?

**Comment by pvanbuijtene** [comment]
👋 @anGie44 the conflicts are resolved.

File: aws/resource_aws_wafv2_web_acl.go [hunk 10]
```diff
@@ -0,0 +1,843 @@
...
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.UpdateWebACL(u)
+	}
+
+	if err != nil {
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl.go:274]: if the error here is a `WAFOptimisticLockException`, AWS docs recommend: 
```
Get the resource again, make any changes you need to make to the new copy, and retry your operation.
```
so I'm wondering as a practitioner, is that apparent if the error were returned to you here? ie. run terraform `plan` to get the new `LockToken` into state and then `apply` again to make the updates, not just retry an `apply` of the plan already created. i say this to think if we need to provide some additional error logging

File: aws/resource_aws_wafv2_web_acl.go [hunk 11]
```diff
@@ -0,0 +1,843 @@
...
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.DeleteWebACL(r)
+	}
+
+	if err != nil {
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl.go:318]: similar comment to the one above

File: aws/resource_aws_wafv2_web_acl.go [hunk 12]
```diff
@@ -0,0 +1,843 @@
...
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				// Required field but currently only supports "IP"
+				"aggregate_key_type": {
+					Type:     schema.TypeString,
+					Optional: true,
+					Default:  wafv2.RateBasedStatementAggregateKeyTypeIp,
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl.go:400]: if required per AWS docs, we could alternatively remove the default but validate this field with a `ValidateFunc: validation.StringInSlice()` in case AWS later adds different aggregate types

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 13]
```diff
@@ -0,0 +1,1120 @@
...
+	"github.com/aws/aws-sdk-go/service/wafv2"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func TestAccAwsWafv2WebACL_Basic(t *testing.T) {
+	var v wafv2.WebACL
+	webACLName := fmt.Sprintf("web-acl-%s", acctest.RandString(5))
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:18]: Nit: by convention we tend to use `rName := acctest.RandomWithPrefix("tf-acc-test")`

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 14]
```diff
@@ -0,0 +1,1120 @@
...
+				ImportStateVerify: true,
+				ImportStateIdFunc: testAccAwsWafv2WebACLImportStateIdFunc(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2WebACL_ChangeNameForceNew(t *testing.T) {
+	var before, after wafv2.WebACL
+	webACLName := fmt.Sprintf("web-acl-%s", acctest.RandString(5))
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:99]: similar comment as above

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 15]
```diff
@@ -0,0 +1,1120 @@
...
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2WebACL_ManagedRuleGroupStatement(t *testing.T) {
+	var v wafv2.WebACL
+	webACLName := fmt.Sprintf("web-acl-%s", acctest.RandString(5))
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:172]: same Nit as above

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 16]
```diff
@@ -0,0 +1,1120 @@
...
+				ImportStateVerify: true,
+				ImportStateIdFunc: testAccAwsWafv2WebACLImportStateIdFunc(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2WebACL_Minimal(t *testing.T) {
+	var v wafv2.WebACL
+	webACLName := fmt.Sprintf("web-acl-%s", acctest.RandString(5))
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:232]: same Nit as above

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 17]
```diff
@@ -0,0 +1,1120 @@
...
+					resource.TestCheckResourceAttr(resourceName, "visibility_config.0.sampled_requests_enabled", "false"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2WebACL_RateBasedStatement(t *testing.T) {
+	var v wafv2.WebACL
+	webACLName := fmt.Sprintf("web-acl-%s", acctest.RandString(5))
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:264]: same Nit as above

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 18]
```diff
@@ -0,0 +1,1120 @@
...
+				ImportStateIdFunc: testAccAwsWafv2WebACLImportStateIdFunc(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2WebACL_RuleGroupReferenceStatement(t *testing.T) {
+	var v wafv2.WebACL
+	var idx int
+	webACLName := fmt.Sprintf("web-acl-%s", acctest.RandString(5))
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:327]: same Nit as above

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 19]
```diff
@@ -0,0 +1,1120 @@
...
+				ImportStateVerify: true,
+				ImportStateIdFunc: testAccAwsWafv2WebACLImportStateIdFunc(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccAwsWafv2WebACL_Tags(t *testing.T) {
+	var v wafv2.WebACL
+	webACLName := fmt.Sprintf("web-acl-%s", acctest.RandString(5))
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl_test.go:393]: same Nit above

File: aws/resource_aws_wafv2_web_acl.go [hunk 20]
```diff
@@ -0,0 +1,843 @@
...
+
+	if v, ok := d.GetOk("description"); ok {
+		params.Description = aws.String(v.(string))
+	}
+
+	if v := d.Get("tags").(map[string]interface{}); len(v) > 0 {
+		params.Tags = keyvaluetags.New(v).IgnoreAws().Wafv2Tags()
+	}
+
+	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
```
Comment (gdavison) [inline aws/resource_aws_wafv2_web_acl.go:156]: We've started creating constants for timeout values, for example `wafv2WebACLCreateTimeout` rather than using the values inline. Please update this and other timeouts.

File: aws/resource_aws_wafv2_web_acl.go [hunk 21]
```diff
@@ -0,0 +1,843 @@
...
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.CreateWebACL(params)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error creating WAFv2 WebACL: %s", err)
```
Comment (gdavison) [inline aws/resource_aws_wafv2_web_acl.go:173]: We prefer wrapping the original error, using the new [error wrapping in Go 1.13](https://blog.golang.org/go1.13-errors#TOC_3.3.), rather than just the string value of the original error. Please also update other error messages

File: aws/resource_aws_wafv2_web_acl_test.go [hunk 22]
```diff
@@ -0,0 +1,1120 @@
...
+    allow {}
+  }
+
+  visibility_config {
+    cloudwatch_metrics_enabled = false
+    metric_name                = "friendly-metric-name"
+    sampled_requests_enabled   = false
+  }
+}
+`, name, name)
```
Comment (gdavison) [inline aws/resource_aws_wafv2_web_acl_test.go:577]: Since the parameters are repeated, could we use argument indexes, i.e. `%[1]s` in `fmt.Sprintf()` instead?

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 23]
```diff
@@ -0,0 +1,495 @@
...
+* `type` - (Required) The transformation to apply, you can specify the following types: `NONE`, `COMPRESS_WHITE_SPACE`, `HTML_ENTITY_DECODE`, `LOWERCASE`, `CMD_LINE`, `URL_DECODE`. See the [documentation](https://docs.aws.amazon.com/waf/latest/APIReference/API_TextTransformation.html) for more details.
+
+### Visibility Configuration
+
+The `visibility_config` block supports the following arguments:
+
+* `cloudwatch_metrics_enabled` - (Required) A boolean indicating whether the associated resource sends metrics to CloudWatch. For the list of available metrics, see [AWS WAF Metrics](https://docs.aws.amazon.com/waf/latest/developerguide/monitoring-cloudwatch.html#waf-metrics).
+* `metric_name` - (Required) A friendly name of the CloudWatch metric. The name can contain only alphanumeric characters (A-Z, a-z, 0-9) hyphen(-) and underscore (_), with length from one to 128 characters. It can't contain whitespace or metric names reserved for AWS WAF, for example `All` and `Default_Action`.
```
Comment (gdavison) [inline website/docs/r/wafv2_web_acl.html.markdown:478]: Some Markdown parsers don't special-case a single underscore character, so italicize the remainder of the document. This can be escaped by using `\_` instead

File: aws/config.go [hunk 24]
```diff
@@ -720,7 +720,7 @@ func (c *Config) Client() (interface{}, error) {
 			r.Retryable = aws.Bool(true)
 		}
 
-		if r.Operation.Name == "CreateIPSet" || r.Operation.Name == "CreateRegexPatternSet" || r.Operation.Name == "CreateRuleGroup" {
+		if r.Operation.Name == "CreateIPSet" || r.Operation.Name == "CreateRegexPatternSet" || r.Operation.Name == "CreateRuleGroup" || r.Operation.Name == "CreateWebACL" {
```
[the remark this answers is not in the document]
Comment (pvanbuijtene) [reply aws/config.go:723]: was thinking the same, but the comment above is longer... consider it done.

File: aws/resource_aws_wafv2_web_acl.go [hunk 25]
```diff
@@ -0,0 +1,843 @@
...
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error deleting WAFv2 WebACL: %s", err)
+	}
+
+	return nil
+}
+
+func wafv2WebACLRootStatementSchema(level int) *schema.Schema {
```
Comment (anGie44) [inline aws/resource_aws_wafv2_web_acl.go:325]: @pvanbuijtene latest discussion in the RuleGroup resource issue #1175 revealed the `statement` field in the RuleGroup resource doesn't include `managed_rules` as included in this schema definition; this shouldn't block the release of this PR but just noting here that this function may later need to be renamed to `wafv2RootStatementSchema` and replace the current function in the wafv2_helper file as the `statement` field in both resources look identical per AWS docs

File: aws/resource_aws_wafv2_web_acl.go (hunk 10 above)
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl.go:274]: @anGie44 it can probably be improved indeed, but if this is done through logging would this message be visible to the end-user? I thought logging should be enabled.

I'm open to any (code) suggestion.

File: aws/resource_aws_wafv2_web_acl.go (hunk 25 above)
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl.go:325]: @anGie44 AFAIK managed rules can only be added on a Web ACL, the docs are not 100% correct 😬 ... this is also why I had to create a different root statement function.

File: aws/resource_aws_wafv2_web_acl.go (hunk 10 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_web_acl.go:274]: oh by "logging" i think we can follow the same `fmt.Errorf` convention but catch this specific case with:
```
	if isAWSErr(err, wafv2. WAFOptimisticLockException, "") {
            return fmt.Errorf("Error updating WAFv2 WebACL, resource has changed since last refresh: %w", err)
	}
       return fmt.Errorf("Error updating WAFv2 WebACL: %s", err)
```

File: aws/resource_aws_wafv2_web_acl.go (hunk 25 above)
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl.go:325]: Sadly I encountered more undocumented limitations like this while working on these wafv2 resources

File: aws/resource_aws_wafv2_web_acl.go (hunk 10 above)
Comment (pvanbuijtene) [reply aws/resource_aws_wafv2_web_acl.go:280]: aaah ok, clear :) ... tried to make it as  clear as possible

File: aws/resource_aws_wafv2_web_acl.go (hunk 25 above)
Comment (anGie44) [reply aws/resource_aws_wafv2_web_acl.go:334]: ah ok I had a feeling there was some backgrounf behind the design decision (and looking with AWS Console I do only see the managed rule group option within the Web ACL UI)

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 26]
```diff
@@ -0,0 +1,495 @@
+---
+subcategory: "WAFv2"
+layout: "aws"
+page_title: "AWS: aws_wafv2_web_acl"
+description: |-
+  Creates a WAFv2 Web ACL resource.
+---
+
+# Resource: aws_wafv2_rule_group
```
Comment (anGie44) [inline website/docs/r/wafv2_web_acl.html.markdown:9]:
```suggestion
# Resource: aws_wafv2_web_acl
```

Ahh just caught this one when running the website docs locally @pvanbuijtene

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 27]
```diff
@@ -0,0 +1,495 @@
...
+
+* `cloudwatch_metrics_enabled` - (Required) A boolean indicating whether the associated resource sends metrics to CloudWatch. For the list of available metrics, see [AWS WAF Metrics](https://docs.aws.amazon.com/waf/latest/developerguide/monitoring-cloudwatch.html#waf-metrics).
+* `metric_name` - (Required) A friendly name of the CloudWatch metric. The name can contain only alphanumeric characters (A-Z, a-z, 0-9) hyphen(-) and underscore (\_), with length from one to 128 characters. It can't contain whitespace or metric names reserved for AWS WAF, for example `All` and `Default_Action`.
+* `sampled_requests_enabled` - (Required) A boolean indicating whether AWS WAF should store a sampling of the web requests that match the rules. You can view the sampled requests through the AWS WAF console.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - The ARN of the WAF rule group.
```
Comment (anGie44) [inline website/docs/r/wafv2_web_acl.html.markdown:485]:
```suggestion
* `arn` - The ARN of the WAF WebACL.
```

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 28]
```diff
@@ -0,0 +1,495 @@
...
+* `capacity` - The web ACL capacity units (WCUs) currently being used by this web ACL.
+* `id` - The ID of the WAF rule group.
```
Comment (anGie44) [inline website/docs/r/wafv2_web_acl.html.markdown:487]:
```suggestion
* `id` - The ID of the WAF WebACL.
```

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 29]
```diff
@@ -0,0 +1,495 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `default_action` - (Required) The action to perform if none of the `rules` contained in the WebACL match. See [Default Action](#default-action) below for details.
+* `description` - (Optional) A friendly description of the rule group.
```
Comment (anGie44) [inline website/docs/r/wafv2_web_acl.html.markdown:259]:
```suggestion
* `description` - (Optional) A friendly description of WebACL.
```

File: website/docs/r/wafv2_web_acl.html.markdown [hunk 30]
```diff
@@ -0,0 +1,495 @@
...
+* `name` - (Required) A friendly name of the rule group.
```
Comment (anGie44) [inline website/docs/r/wafv2_web_acl.html.markdown:260]:
```suggestion
* `name` - (Required) A friendly name of the WebACL.
```

**Comment by pvanbuijtene** [comment]
@anGie44 I see I was getting a bit sloppy at the end 😬 

Had to fix a rebase issue, maybe wise to run the integration tests again... just to be sure

**Comment by anGie44** [comment]
no worries @pvanbuijtene, thanks for taking a look! will comment back with AccTest results
